### PR TITLE
[CWS] Add ptracer

### PR DESCRIPTION
**Pull request by spikat**
### What does this PR do?

Add the new ptracer to the agent, and embed it in the cws-instrumentation bin/image. Also:
- Fix (most of) linters issues
- ARM64 port (to be tested)
- forward signals to traced processes
- handle dup/dup2/dup3 to maintain fd/path cache
- handle cwd to return only absolute paths
- handle chdir/fchdir to maintain cwd cache right

### Motivation

Support fargate

### Describe how to test/QA your changes

Here is some test funcs to validate the features (to be integrate as tests!! :))

```

> #define _GNU_SOURCE
> #include <stdlib.h>
> #include <unistd.h>
> #include <stdio.h>
> #include <signal.h>
> #include <fcntl.h>
> #include <sys/types.h>
> #include <sys/stat.h>
> 
> int mychdir(char* path) {
>     int fd = open(path, O_RDONLY);
>     if (fd < 0) {
>         perror("mychdir open");
>         return (-1);
>     }
>     int newfd = dup(fd);
>     return (fchdir(newfd));
> }
> int main(int ac, char** av) {
>     int fd = open("1/foo", O_RDWR);
>     if (fd < 0) {
>         perror("open 1");
>         exit(1);
>     } else { close(fd); }
>     fd = open("1/2/foo", O_RDWR);
>     if (fd < 0) {
>         perror("open 2");
>         exit(1);
>     } else { close(fd); }
>     if (mychdir("1")) {
>         perror("chdir");
>     }
>     char* cwd = getcwd(NULL, 0);
>     printf("CWD: %s\n", cwd);
>     free(cwd);
>     fd = open("foo", O_RDWR);
>     if (fd < 0) {
>         perror("open 3");
>         exit(1);
>     } else { close(fd); }
>     if (mychdir("2")) {
>         perror("chdir");
>     }
>     cwd = getcwd(NULL, 0);
>     printf("CWD: %s\n", cwd);
>     free(cwd);
>     fd = open("foo", O_RDWR);
>     if (fd < 0) {
>         perror("open 4");
>         exit(1);
>     } else { close(fd); }
>     if (mychdir("3")) {
>         perror("chdir");
>     }
>     cwd = getcwd(NULL, 0);
>     printf("CWD: %s\n", cwd);
>     free(cwd);
>     fd = open("foo", O_RDWR);
>     if (fd < 0) {
>         perror("open 5");
>         exit(1);
>     } else { close(fd); }
>     return (0);
> }
> 
> int main(int ac, char** av) {
>     int fd = open("1/foo", O_RDWR);
>     if (fd < 0) {
>         perror("open 1");
>         exit(1);
>     } else { close(fd); }
>     fd = open("1/2/foo", O_RDWR);
>     if (fd < 0) {
>         perror("open 2");
>         exit(1);
>     } else { close(fd); }
>     if (chdir("1")) {
>         perror("chdir");
>     }
>     char* cwd = getcwd(NULL, 0);
>     printf("CWD: %s\n", cwd);
>     free(cwd);
>     fd = open("foo", O_RDWR);
>     if (fd < 0) {
>         perror("open 3");
>         exit(1);
>     } else { close(fd); }
>     if (chdir("2")) {
>         perror("chdir");
>     }
>     cwd = getcwd(NULL, 0);
>     printf("CWD: %s\n", cwd);
>     free(cwd);
>     fd = open("foo", O_RDWR);
>     if (fd < 0) {
>         perror("open 4");
>         exit(1);
>     } else { close(fd); }
>     if (chdir("3")) {
>         perror("chdir");
>     }
>     cwd = getcwd(NULL, 0);
>     printf("CWD: %s\n", cwd);
>     free(cwd);
>     fd = open("foo", O_RDWR);
>     if (fd < 0) {
>         perror("open 5");
>         exit(1);
>     } else { close(fd); }
>     return (0);
> }
> 
> int main(int ac, char** av) {
>     /* int fd = open("./bar", O_CREAT|O_TRUNC|O_RDWR, S_IWUSR|S_IRUSR|S_IRGRP|S_IROTH); */
>     int fd = open("./bar", O_RDWR);
>     if (fd == -1) {
>         perror("open");
>         return (-1);
>     }
>     dprintf(1, "fd %i\n", fd);
>     dprintf(fd, "fd %i\n", fd);
>     int _dup = dup(fd);
>     if (_dup == -1) {
>         fprintf(stderr, "dup error\n");
>         return (-1);
>     }
>     dprintf(1, "dup %i\n", _dup);
>     dprintf(_dup, "dup %i\n", _dup);
>     int _dup2 = dup2(_dup, 42);
>     if (_dup2 == -1) {
>         fprintf(stderr, "dup2 error\n");
>         return (-1);
>     }
>     dprintf(1, "dup2 %i\n", _dup2);
>     dprintf(_dup2, "dup2 %i\n", _dup2);
>     int _dup3 = dup3(_dup2, 51, 0);
>     if (_dup3 == -1) {
>         fprintf(stderr, "dup3 error\n");
>         return (-1);
>     }
>     dprintf(1, "dup3 %i\n", _dup3);
>     dprintf(_dup3, "dup3 %i\n", _dup3);
>     close(fd);
>     close(_dup);
>     close(_dup2);
>     close(_dup3);
>     return (0);
> }
> 
> void sighandler(int signum) {
>     printf("!! SIGHANDLER !! %i\n", signum);
> }
> int main(int ac, char **av) {
>     int sigs[] = { SIGUSR1, SIGUSR2, SIGTERM, SIGHUP, 0 };
>     for (int i = 0; sigs[i] != 0; i++) {
>         if (signal(sigs[i], sighandler) == SIG_ERR) {
>             printf("Error registering the signal %i\n", sigs[i]);
>             exit(1);
>         }
>         printf("SIGNAL %i registered\n", sigs[i]);
>     }
>     while (42) {
>         pause();
>     }
>     return (0);
> }
> 

```

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by safchain — APPROVED** [review]

**Review by paulcacheux — APPROVED** [review]

File: pkg/security/ptracer/ptracer.go [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+func (t *Tracer) ReadArgInt64(regs syscall.PtraceRegs, arg int) int64 {
+	return int64(t.argToRegValue(regs, arg))
+}
+
+// ReadArgInt32 reads the and returns the wanted arg as int32
+func (t *Tracer) ReadArgInt32(regs syscall.PtraceRegs, arg int) int32 {
+	return int32(t.argToRegValue(regs, arg))
+}
+
+// ReadArgUint32 reads the and returns the wanted arg as uint32
```
Comment (lebauce) [inline pkg/security/ptracer/ptracer.go:126]: nit: missing word in these comments

File: pkg/security/ptracer/ptracer.go [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+		i      uint64
+	)
+
+	for {
+		n, err := syscall.PtracePeekData(pid, uintptr(ptr+i), data)
+		if err != nil || n != len(data) {
+			return result, err
+		}
+
+		ptr := binary.LittleEndian.Uint64(data)
```
Comment (lebauce) [inline pkg/security/ptracer/ptracer.go:158]: Isn't it platform dependent ?

File: pkg/security/ptracer/ptracer.go [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+
+			if err := syscall.PtraceCont(pid, 0); err != nil {
+				continue
+			}
+		}
+	}
+
+	return nil
+}
+func traceFilterProg(opts Opts) (*syscall.SockFprog, error) {
```
Comment (lebauce) [inline pkg/security/ptracer/ptracer.go:257]: nit: empty line

File: pkg/security/ptracer/syscalls_arm64.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+package ptracer
+
+import (
+	"syscall"
+
+	"golang.org/x/sys/unix"
+)
+
+const (
+	OpenatNr   = 56  // OpenatNr defines the syscall ID for arm64
```
Comment (lebauce) [inline pkg/security/ptracer/syscalls_arm64.go:17]: shouldn't openat2 be part of this list ?

File: cmd/cws-instrumentation/subcommands/tracecmd/trace.go [hunk 5]
```diff
@@ -0,0 +1,608 @@
...
+	Cwd string
+}
+
+func fillProcessCwd(process *Process) error {
+	cwd, err := os.Readlink(fmt.Sprintf("/proc/%d/cwd", process.Pid))
+	if err != nil {
+		return err
+	}
+	process.Cwd = cwd
+	log.Debugf("fillProcessCwd: %v", cwd)
```
Comment (lebauce) [inline cmd/cws-instrumentation/subcommands/tracecmd/trace.go:78]: nit: is it intended ?

**Review by YoannGh — APPROVED** [review]

File: pkg/security/ptracer/ptracer.go (hunk 2 above)
Comment (safchain) [reply pkg/security/ptracer/ptracer.go:158]: Fixed

File: pkg/security/ptracer/syscalls_arm64.go (hunk 4 above)
Comment (safchain) [reply pkg/security/ptracer/syscalls_arm64.go:17]: nice catch, thanks

File: pkg/security/ptracer/syscalls_amd64.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+		syscall.PTRACE_O_TRACEEXEC |
+		syscall.PTRACE_O_TRACESYSGOOD |
+		unix.PTRACE_O_TRACESECCOMP
+)
+
+var (
+	// PtracedSyscalls defines the list of syscall we want to ptrace
+	PtracedSyscalls = []string{
+		"open",
+		"openat",
```
Comment (safchain) [inline pkg/security/ptracer/syscalls_amd64.go:46]: add openat2

**Pull request merged**